### PR TITLE
Replace deprecated IndexFormatter

### PR DESCRIPTION
**Pull request by stephenworsley**
### Description

Addresses #3848 (and the duplicate issue #3821).

The replacement formatter is not designed to be an exact copy of IndexFormatter since it seems that IndexFormatter can be considered bugged. Specifically, IndexFormatter would add a duplicate of the label at the 0 tick to the -1 tick. This behaviour was being accounted for by our tests, which have been changed to reflect the new behaviour. The new behaviour also relies on ticks being located precisely on integers (the IndexFormatter would perform some rudimentary rounding operations). This should always be the case since the locator is `MaxNLocator(integer=True)`.

### CODE REVIEW

File: lib/iris/plot.py [hunk 1]
```diff
@@ -249,7 +249,19 @@ def _string_coord_axis_tick_labels(string_axes, axes=None):
 
     ax = axes if axes else plt.gca()
     for axis, ticks in string_axes.items():
-        formatter = mpl_ticker.IndexFormatter(ticks)
+        # Define a tick formatter, this replaces the now deprecated
+        # mpl_ticker.IndexFormatter. Note that this formatter behaves
+        # differently to IndexFormatter and will only assign labels to
```
Comment (trexfeathers) [inline lib/iris/plot.py:254]: I don't think it's right to talk about a change in behaviour in a comment - that's only appropriate for the PR and the "What's New". Do you have some simple context on why `FuncFormatter` needs an input function like this?

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Thanks @stephenworsley ❤️ 

I'm satisfied that:
* We already have enough test coverage to allow for the new behaviour
  (since you had to change an expected result as part of this change, for a valid reason)
* Nothing has broken
  (since all the tests pass)

Please could you address my one comment, and I think we'll be done 🙂

File: lib/iris/plot.py (hunk 1 above)
Comment (stephenworsley) [reply lib/iris/plot.py]: I've tried to clarify what the function is doing by making some of the variables more descriptive. I'm hoping that this + the new comment should make it clear how FuncFormatter is using this function. (i.e. that this function takes in the location of a tick and returns a label.) I felt that making it too explicit what was going on would have made the comment a little too verbose.

Comment (trexfeathers) [reply lib/iris/plot.py]: That's super helpful. I had been seeing `FuncFormatter` as a closed magic box but I appreciate much better what's going on now.

**Review by trexfeathers — APPROVED** [review]

**Pull request merged**